src/client/datascience/jupyter/kernels/kernel.ts

```diff
@@ -105,54 +106,24 @@ export class Kernel implements IKernel {
         );
     }
     public async executeCell(cell: NotebookCell): Promise<void> {
-        await this.start({ disableUI: false });
-        await this.kernelExecution.executeCell(cell);
+        const notebookPromise = this.startNotebook({ disableUI: false });
+        await this.kernelExecution.executeCell(notebookPromise, cell);
     }
     public async executeAllCells(document: NotebookDocument): Promise<void> {
-        await this.start({ disableUI: false });
-        await this.kernelExecution.executeAllCells(document);
+        const notebookPromise = this.startNotebook({ disableUI: false });
+        await this.kernelExecution.executeAllCells(notebookPromise, document);
     }
     public async cancelCell(cell: NotebookCell) {
         this.startCancellation.cancel();
         await this.kernelExecution.cancelCell(cell);
     }
     public async cancelAllCells(document: NotebookDocument) {
         this.startCancellation.cancel();
-        this.kernelExecution.cancelAllCells(document);
+        await this.kernelExecution.cancelAllCells(document);
     }
     public async start(options?: { disableUI?: boolean }): Promise<void> {
-        if (this.restarting) {
-            await this.restarting.promise;
-        }
-        if (this._notebookPromise) {
-            await this._notebookPromise;
-            return;
-        } else {
-            this.startCancellation = new CancellationTokenSource();
-            await this.validate(this.uri);
-            this._notebookPromise = this.notebookProvider.getOrCreateNotebook({
-                identity: this.uri,
-                resource: this.uri,
-                disableUI: options?.disableUI,
-                getOnly: false,
-                metadata: undefined, // No need to pass this, as we have a kernel connection (metadata is required in lower layers to determine the kernel connection).
-                kernelConnection: this.kernelConnectionMetadata,
-                token: this.startCancellation.token
-            });
-
-            this._notebookPromise
-                .then((nb) => (this.kernelExecution.notebook = this.notebook = nb))
```

Comment:
Removed the hack of intiailizing `kernelExecution.notebook`.
Now we pass this aa an argument to `executeCell` which is a better way than initialing a property & then calling a method (& then in the method checking if the property was initialized - yuck)